dali/kernels/shape.h

```diff
@@ -0,0 +1,609 @@
+  assert(0 <= other_ndim && other_ndim <= get_sample_dim() &&
+         "Number of dimensions in subshape must be between 0 and sample_dim()");
+  TensorListShape<other_ndim> result;
+  result.shapes.resize(other_ndim * get_size());
+  int start_offset = get_sample_dim() - other_ndim;
+  for (int sample = 0; sample < get_size(); sample++) {
+    for (int d = 0; d < other_ndim; d++) {
+      result.shapes[sample * other_ndim + d] = shapes[sample * get_sample_dim() + start_offset + d];
+    }
+  }
+  return result;
+}
+
+template <typename Derived, int sample_ndim>
+TensorListShape<DynamicDimensions> TensorListShapeBase<Derived, sample_ndim>::first(int count) {
+  assert(0 <= count && count <= get_sample_dim() &&
+         "Number of dimensions in subshape must be between 0 and sample_dim()");
+  TensorListShape<DynamicDimensions> result;
+  result.shapes.resize(count * get_size());
+  for (int sample = 0; sample < get_size(); sample++) {
+    for (int d = 0; d < count; d++) {
+      result.shapes[sample * count + d] = shapes[sample * get_sample_dim() + d];
+    }
+  }
+  return result;
+}
+
+template <typename Derived, int sample_ndim>
+TensorListShape<DynamicDimensions> TensorListShapeBase<Derived, sample_ndim>::last(int count) {
+  assert(0 <= count && count <= get_sample_dim() &&
+         "Number of dimensions in subshape must be between 0 and sample_dim()");
+  TensorListShape<DynamicDimensions> result;
+  result.shapes.resize(count * get_size());
+  int start_offset = get_sample_dim() - count;
+  for (int sample = 0; sample < get_size(); sample++) {
+    for (int d = 0; d < count; d++) {
+      result.shapes[sample * count + d] = shapes[sample * get_sample_dim() + start_offset + d];
+    }
+  }
+  return result;
+}
+
+template <int left_ndim, int right_ndim>
+bool operator==(const TensorListShape<left_ndim> &left, const TensorListShape<right_ndim> &right) {
+  if (left.sample_dim() != right.sample_dim()) {
+    return false;
+  }
+  if (left.size() != right.size()) {
+    return false;
+  }
+  return left.shapes == right.shapes;
+}
+
+template <int left_ndim, int right_ndim>
+bool operator!=(const TensorListShape<left_ndim> &left, const TensorListShape<right_ndim> &right) {
+  return !(left == right);
+}
+
+template <int sample_ndim>
+std::vector<ptrdiff_t> calculate_offsets(const TensorListShape<sample_ndim> &tls) {
+  std::vector<ptrdiff_t> offsets;
+  offsets.resize(tls.size() + 1);
+  offsets[0] = 0;
+  for (int i = 0; i < tls.size(); i++) {
+    auto sample_shape_span = tls.tensor_shape_span(i);
+    offsets[i + 1] = offsets[i] + volume(sample_shape_span);
+  }
+  return offsets;
+}
+
+template <int ndim>
+bool is_uniform(const TensorListShape<ndim> &tls) {
+  if (!tls.size()) {
+    return true;  // empty is uniform
+  }
+  auto first_span = tls.tensor_shape_span(0);
+  for (int i = 1; i < tls.size(); i++) {
+    if (first_span != tls.tensor_shape_span(i)) {
```

Comment:
Our current span doesn't implement comparison.

Reply:
Actually it is supposed to have: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0122r7.pdf page 11.
I will add it, as it will be useful.